content/en/blog/_posts/2023-07-18-swap-beta1-graduation/index.md

```diff
@@ -0,0 +1,231 @@
+usage. This issue has been addressed in the Beta version introduced in Kubernetes 1.28, which now
+provides the capability to monitor swap usage through several different methods.
+
+The beta version of kubelet now collects
+[node-level metric statistics](/docs/reference/instrumentation/node-metrics/),
+which can be accessed at the /metrics/resource and /stats/summary endpoints. This allows users to
+monitor swap usage and remaining swap memory when using LimitedSwap. Additionally, a
+machine_swap_bytes metric has been added to cadvisor to show the total physical swap capacity of the
+machine.
+
+## Caveats
+Having swap available on a system reduces predictability. Swap's performance is
+worse than regular memory, sometimes by many orders of magnitude, which can
+cause unexpected performance regressions. Furthermore, swap changes a system's
+behaviour under memory pressure, and applications cannot directly control what
+portions of their memory usage are swapped out. Since enabling swap permits
+greater memory usage for workloads in Kubernetes that cannot be predictably
+accounted for, it also increases the risk of noisy neighbours and unexpected
+packing configurations, as the scheduler cannot account for swap memory usage.
+
+The performance of a node with swap memory enabled depends on the underlying
+physical storage. When swap memory is in use, performance will be significantly
+worse in an I/O operations per second (IOPS) constrained environment, such as a
+cloud VM with I/O throttling, when compared to faster storage mediums like
+solid-state drives or NVMe.
+
+As such, we do not advocate the utilization of swap memory for workloads or
+environments that are subject to performance constraints. Furthermore, it is
+recommended to employ `LimitedSwap`, as this significantly mitigates the risks
+posed to the node.
+
+Cluster administrators and developers should benchmark their nodes and applications
+before using swap in production scenarios, and [we need your help](#how-do-i-get-involved) with that!
+
+### Security risk
+Enabling swap on a system without encryption poses a security risk, as critical information,
+such as Kubernetes secrets, may be swapped out to the disk. If an unauthorized individual gains
+access to the disk, they could potentially obtain these secrets. To mitigate this risk, it is
+recommended to use encrypted swap. However, handling encrypted swap is not within the scope of
+kubelet; rather, it is a general OS configuration concern and should be addressed at that level.
+It is the administrator's responsibility to provision encrypted swap to mitigate this risk.
+
+Furthermore, as previously mentioned, with `LimitedSwap` the user has the option to completely
+disable swap usage for a container by specifying memory requests that are equal to memory limits.
+This will prevent the corresponding containers from accessing swap memory and eliminate any
+associated risks of information exposure.
```

Comment:
```suggestion
Does this prevent Pod-level data such as tmpfs volumes getting written to swap? How sure are we of that?
```

Reply:
So currently we only deal with swap on a container-level basis, not a pod-level basis.

Let's say that you create a non-Burstable pod, or a pod which its memory limits equal memory requests. From a cgroup perspective, the containers' `memory.swap.max` will be set to 0. The pod's `memory.swap.max` is not configured by kubelet.

IIUC, this is completely fine. In cgroup v2, processes might only reside at a leaf cgroup, so all of the processes would run one of the containers' cgroup and wouldn't be able to swap its memory. I'm not even sure if setting  `memory.swap.max` to 0 on the pod level cgroup would have any effect.

In addition, investigating tmpfs volumes is [one of the requirements to Beta2](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#beta-2).

Reply:
> In addition, investigating tmpfs volumes is [one of the requirements to Beta2](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#beta-2).

We need to document something for security for beta, even if the advice is “The Kubernetes project has not evaluated whether this feature is secure”.

Bear in mind that Secret data in a _tmpfs_ doesn't belong to any Pod process, not even the sandbox.

Reply:
The KEP [addresses these security risks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#security-risk) and says:
> Enabling swap on a system without encryption poses a security risk, as critical information, such as Kubernetes secrets, may be swapped out to the disk. If an unauthorized individual gains access to the disk, they could potentially obtain these secrets. To mitigate this risk, it is recommended to use encrypted swap. However, handling encrypted swap is not within the scope of kubelet; rather, it is a general OS configuration concern and should be addressed at that level. Nevertheless, it is essential to provide documentation that warns users of this potential issue, ensuring they are aware of the potential security implications and can take appropriate steps to safeguard their system.

As the KEP dictates, the documentation provides a warning about the potential risks of secrets being paged to the disk. In addition, as the KEP dictates, we encourage users to use encrypted swap to mitigate this risk.

Furthermore, the behavior I've outlined before (pod's cgroup having `memory.swap.max == max`) isn't change with the recent changes to swap. IOW, it's the same behavior as in Alpha.

As said, this risk is mitigated by using encrypted swap.

In any case, I've tried to reword this to improve the phrasing.

Reply:
Don't know if we want to provide a specific link to the secrets docs page which mentions that secrets are held in a `tmpfs` and not written to disk usually (without swap) https://kubernetes.io/docs/concepts/configuration/secret/#:~:text=For%20mounting%20secrets%20into%20Pods,confidential%20data%20from%20the%20Secret. 

This might help make the point clearer about why encrypted swap is recommended.

Reply:
@raesene Done, thanks!
PTAL